dartagnan/src/main/java/com/dat3m/dartagnan/expression/utils/ExpressionHelper.java

```diff
@@ -13,15 +17,56 @@ public static void checkSameType(Expression x, Expression y) {
                 "The types of %s and %s do not match.", x, y);
     }
 
+    public static void checkSameType(Type x, Type y) {
+        Preconditions.checkArgument(x.equals(y),
+                "The types %s and %s do not match.", x, y);
+    }
+
     public static void checkExpectedType(Expression expr, Class<? extends Type> expectedClass) {
         final Class<? extends Type> exprClass = expr.getType().getClass();
         Preconditions.checkArgument(expectedClass.isAssignableFrom(exprClass),
-                "The expression '%s' has unexpected type class '%s', expected type class is '%s'",
+                "The expression '%s' has unexpected type class '%s', expected type class is '%s'.",
                 expr, exprClass.getSimpleName(), expectedClass.getSimpleName());
     }
 
     public static void checkSameExpectedType(Expression x, Expression y, Class<? extends Type> expectedClass) {
         checkSameType(x, y);
         checkExpectedType(x, expectedClass);
     }
+
+    public static void checkInbounds(Type aggregateType, int index) {
+        Preconditions.checkArgument(isAggregateLike(aggregateType), "Non-aggregate type %s.", aggregateType);
+        Preconditions.checkArgument(index >= 0, "Negative index: %s.", index);
+        if (aggregateType instanceof AggregateType aggType) {
+            checkArgument(index < aggType.getFields().size(),
+                    "Index %s out of bounds for type %s.",
+                    index, aggregateType);
+        } else if (aggregateType instanceof ArrayType arrayType) {
+            checkArgument(!arrayType.hasKnownNumElements() || index < arrayType.getNumElements(),
+                    "Index %s out of bounds for type %s.",
+                    index, arrayType);
+        }
+    }
+
+    public static boolean isAggregateLike(Type type) {
+        return type instanceof AggregateType || type instanceof ArrayType;
+    }
+
+    public static boolean isAggregateLike(Expression expr) {
+        return isAggregateLike(expr.getType());
+    }
+
+    public static Type extractType(Type type, Iterable<Integer> indices) {
+        for (int index : indices) {
+            checkInbounds(type, index);
+            if (type instanceof AggregateType aggregateType) {
+                type = aggregateType.getFields().get(index).type();
+            } else if (type instanceof ArrayType arrayType) {
+                type = arrayType.getElementType();
+            } else {
+                assert false : "unreachable";
```

Comment:
We should avoid this kind of assertions. As I mentioned recently in a previous PR, assertions are disabled once we build the artifact and thus, they do not get triggered. If the check is important, we should throw an exception

Reply:
And what if the check is unimportant? The code is literally unreachable, but I didn't wanna leave out the else case without documentation. The assertion is just a better version of a "// unreachable" comment.
Anyways, I can upgrade the assertion to a proper exception, even if it cannot ever trigger.